File: probs/33.cpp
#include <vector>
using namespace std;

int search(vector<int> &nums, int target)
{
    int pivot = find_pivot(nums);
    if (pivot == -1)
    {
        return binary_search(nums, 0, nums.size() - 1, target);
    }

    if (target == nums[pivot])
    {
        return pivot;
    }

    if (nums[0] <= target)
    {
        return binary_search(nums, 0, pivot - 1, target);
    }

    return binary_search(nums, pivot, nums.size() - 1, target);
}

int binary_search(vector<int> &nums, int lo, int hi, int key)
{
    if (lo > hi)
    {
        return -1;
    }

    int mid = lo + (hi - lo) / 2;
    if (nums[mid] == key)
    {
        return mid;
    }

    if (nums[mid] < key)
    {
        return binary_search(nums, mid + 1, hi, key);
    }

    return binary_search(nums, lo, mid - 1, key);
}

int find_pivot(vector<int> &nums)
{
    int lo = 0;
    int hi = nums.size() - 1;

    while (lo < hi)
    {
        int mid = lo + (hi - lo) / 2;
        if (mid < hi && nums[mid + 1] < nums[mid])
        {
            return mid + 1;
        }

        if (mid > lo && nums[mid - 1] > nums[mid])
        {
            return mid;
        }

        if (nums[lo] >= nums[mid])
        {
            hi = mid - 1;
        }
        else
        {
            lo = mid + 1;
        }
    }

    return -1;
}

File: zig/33.zig
const std = @import("std");

fn solution(nums: []i32, target: i32) i32 {
    var lo: usize = 0;
    var hi: usize = nums.len - 1;

    while (lo <= hi) {
        const mid = lo + @divFloor(hi - lo, 2);
        if (nums[mid] == target) {
            return @as(i32, @intCast(mid));
        }

        // there's a possibility that the left is ordered
        if (nums[mid] >= nums[lo]) {
            // search the left
            if (nums[lo] <= target and target <= nums[mid]) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        } else {
            if (nums[mid] <= target and target <= nums[hi]) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
    }

    return -1;
}

const expect = std.testing.expect;

test "test case #1" {
    var input = [_]i32{ 4, 5, 6, 7, 0, 1, 2 };

    const target = 0;

    const sol = solution(&input, target);

    try expect(sol == 4);
}

test "test case #2" {
    var input = [_]i32{ 4, 5, 6, 7, 0, 1, 2 };

    const target = 3;

    const sol = solution(&input, target);

    try expect(sol == -1);
}

test "test case #3" {
    var input = [_]i32{1};

    const target = 0;

    const sol = solution(&input, target);

    try expect(sol == -1);
}
